crates/solver-service/src/apis/quote/eip712.rs

```diff
@@ -0,0 +1,271 @@
+		}],
+		"spender": format!("0x{:x}", spender),
+		"nonce": nonce_ms.to_string(),
+		"deadline": deadline_secs.to_string(),
+		"witness": {
+			"expires": expires_u32,
+			"inputOracle": format!("0x{:x}", input_oracle),
+			"outputs": [{
+				"oracle": format!("0x{:064x}", 0),
+				"settler": format!("0x{}{:x}", "0".repeat(24), output_settler),
+				"chainId": dest_chain_id,
+				"token": format!("0x{}{:x}", "0".repeat(24), dest_token),
+				"amount": amount.to_string(),
+				"recipient": format!("0x{}{:x}", "0".repeat(24), recipient),
+				"call": "0x",
+				"context": "0x"
+			}]
+		}
+	});
+
+	Ok((format_hex(&final_digest), message_json))
+}
+
+/// Resolve the oracle address for a given chain from settlement implementation config.
+fn resolve_oracle_address(config: &Config, chain_id: u64) -> Result<AlloyAddress, QuoteError> {
+	let Some(impl_val) = config.settlement.implementations.get("eip7683") else {
+		return Err(QuoteError::InvalidRequest(
+			"Missing settlement.implementations.eip7683 in config".to_string(),
+		));
+	};
+	let Some(table) = impl_val.as_table() else {
+		return Err(QuoteError::InvalidRequest(
+			"Invalid eip7683 settlement implementation format".to_string(),
+		));
+	};
+	let Some(oracle_map) = table.get("oracle_addresses").and_then(|v| v.as_table()) else {
+		return Err(QuoteError::InvalidRequest(
+			"Missing oracle_addresses in eip7683 settlement implementation".to_string(),
+		));
+	};
+	let key = chain_id.to_string();
+	let Some(addr_str) = oracle_map.get(&key).and_then(|v| v.as_str()) else {
+		return Err(QuoteError::InvalidRequest(format!(
+			"Oracle address not configured for chain {}",
+			chain_id
+		)));
+	};
+	addr_str
+		.parse::<AlloyAddress>()
+		.map_err(|e| QuoteError::InvalidRequest(format!("Invalid oracle address: {}", e)))
+}
+
+/// Resolve the Permit2 address for a given chain strictly from config.
+/// Looks under `settlement.implementations.eip7683.permit2_addresses`.
+pub fn resolve_permit2_address(config: &Config, chain_id: u64) -> Result<AlloyAddress, QuoteError> {
+	// Use default Permit2 address mapping from CustodyStrategy
+	use super::custody::CustodyStrategy;
+	let map = CustodyStrategy::default_permit2_addresses();
+	let solver_address = map.get(&chain_id).ok_or_else(|| {
+		QuoteError::InvalidRequest(format!("No default Permit2 address for chain {}", chain_id))
+	})?;
+
+	// Convert solver_types::Address to alloy_primitives::Address
+	if solver_address.0.len() != 20 {
+		return Err(QuoteError::InvalidRequest(
+			"Invalid address length".to_string(),
+		));
+	}
+	let mut addr_bytes = [0u8; 20];
+	addr_bytes.copy_from_slice(&solver_address.0);
+	Ok(AlloyAddress::from(addr_bytes))
+}
+
+fn bytes20_to_address(bytes: &[u8]) -> Result<AlloyAddress, QuoteError> {
+	if bytes.len() != 20 {
+		return Err(QuoteError::InvalidRequest(format!(
+			"Expected 20-byte address, got {}",
+			bytes.len()
+		)));
+	}
+	let mut arr = [0u8; 20];
+	arr.copy_from_slice(bytes);
+	Ok(AlloyAddress::from(arr))
+}
```

Comment:
Wasn't this moved to types? This is old code that somehow made its way back here?